lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -139,21 +138,67 @@ withNetworkLayer tr project k = k NetworkLayer
     runBlockfrost =
         ExceptT . (first ClientError <$>) . BF.runBlockfrostClientT project
 
-blockToBlockHeader ::
-    forall m. MonadError BlockfrostError m => BF.Block -> m BlockHeader
-blockToBlockHeader block@BF.Block{..} = do
-    slotNo <- case _blockSlot of
-        Just s -> pure $ SlotNo $ fromIntegral $ BF.unSlot s
-        Nothing -> throwError $ NoSlotError block
-    blockHeight <- case _blockHeight of
-        Just height -> pure $ Quantity $ fromIntegral height
-        Nothing -> throwError $ NoBlockHeight block
-    headerHash <- parseBlockHeader _blockHash
-    parentHeaderHash <- for _blockPreviousBlock parseBlockHeader
-    pure BlockHeader { slotNo, blockHeight, headerHash, parentHeaderHash }
-  where
-    parseBlockHeader :: BF.BlockHash -> m (Hash "BlockHeader")
-    parseBlockHeader blockHash =
-        case fromText (BF.unBlockHash blockHash) of
-            Right hash -> pure hash
-            Left tde -> throwError $ InvalidBlockHash blockHash tde
+class FromBlockfrost b w where
+    fromBlockfrost :: forall m. MonadError BlockfrostError m => b -> m w
+
+instance FromBlockfrost BF.Block BlockHeader where
+    fromBlockfrost block@BF.Block{..} = do
+        slotNo <- case _blockSlot of
+            Just s -> pure $ SlotNo $ fromIntegral $ BF.unSlot s
+            Nothing -> throwError $ NoSlotError block
+        blockHeight <- case _blockHeight of
+            Just height -> pure $ Quantity $ fromIntegral height
+            Nothing -> throwError $ NoBlockHeight block
+        headerHash <- parseBlockHeader _blockHash
+        parentHeaderHash <- for _blockPreviousBlock parseBlockHeader
+        pure BlockHeader { slotNo, blockHeight, headerHash, parentHeaderHash }
+      where
+        parseBlockHeader blockHash =
+            case fromText (BF.unBlockHash blockHash) of
+                Right hash -> pure hash
+                Left tde -> throwError $ InvalidBlockHash blockHash tde
+
+instance FromBlockfrost BF.ProtocolParams ProtocolParameters where
+    fromBlockfrost BF.ProtocolParams{..} = do
+        decentralizationLevel <-
+            let percentage = mkPercentage $
+                    toRational _protocolParamsDecentralisationParam
+            in case percentage of
+                Left PercentageOutOfBoundsError ->
+                    throwError $ InvalidDecentralizationLevelPercentage
+                        _protocolParamsDecentralisationParam
+                Right level -> pure $ DecentralizationLevel level
+        let intToQuantity :: (Num q, Integral i) => i -> Quantity s q
+            intToQuantity = Quantity . fromIntegral
+            txParameters = TxParameters
+                { getFeePolicy =  LinearFee
+                    (intToQuantity _protocolParamsMinFeeA)
+                    (intToQuantity _protocolParamsMinFeeB)
```

Comment:
This should be the other way around 🙈 (I guess ideally `LinearFee` should swap its order to match a-then-b or we should use record field names… this is easy to mess up)

Reply:
wait, I don't get it:
The formula is: `a + b * x`
constant factor is the one that is added once, named `a` and tagged "lovelace"
linear factor is the one that grows linearly with every byte, the multiplier named `b` and tagged "lovelace/byte"

Its Blockfrost API that mixed up documentation (they say A is linear and B is constant but it should be other way around), our code has it correct, agree?

Reply:
So blockfrost has
```
a :: "lovelace/byte"
b :: "lovelace"
```
Which is the terminology used in the ledger, specs, and cardano-api: https://github.com/input-output-hk/cardano-node/blob/73f9a746362695dc2cb63ba757fbcabb81733d23/cardano-api/src/Cardano/Api/ProtocolParameters.hs#L1421-L1422

The wallet however has the constructor
```
LinearFee :: "lovelace" -> "lovelace/byte" -> FeePolicy
```
because of legacy reasons
so you should call it with `LinearFee (b :: "lovelace") (a :: "lovelace/byte")`.

Reply:
Now I see where the confusion comes from! Thanks for taking time explaining it. 
One question:
> but confusingly, a would correspond to minFeeB and b to minFeeA

Should I go on and switch `a` and `b` in our code for the `LinearFee`  to remove confusion going forward, 
or
are those legacy reasons still exist and I'd better not to add even more confusion by trying to un-confuse it? :)

Reply:
I suggest the following: Change the data type to

```
data FeePolicy = LinearFee
    { linearFeeIntercept :: Double -- b
    , linearFeeSlope :: Double -- a
    } deriving (… etc)
```
This follows traditional nomenclature on [linear functions](https://en.wikipedia.org/wiki/Linear_function).

I'm not sure if we should keep the `Quantity "name"` wrappers or not. The main idea behind them is probably that they help prevent this kind of bug ("accidental flip of two arguments") by having different types. Unfortunately, they have not helped in this case, and they do add noise elsewhere. 🤔 I have the feeling that using record fields with traditional nomenclature is more helpful.

Reply:
I like the idea of using traditional nomenclature and agree that tagging didn't help *in this case*.
I'd like to bring @HeinrichApfelmus 's suggestion even further changing the data type design along the lines of Nikita's answer: https://www.reddit.com/r/haskell/comments/3pylic/records_with_different_fields_for_its/cwayig4/?utm_source=reddit&utm_medium=web2x&context=3

```haskell
data LinearFunction = LinearFunction
    { linearFunctionIntercept :: Double
    , linearFunctionSlope :: Double
    }
    deriving (Eq, Show, Generic)
    
newtype FeePolicy = LinearFee LinearFunction
    deriving (Eq, Show, Generic)
```
wdyt?

Reply:
I agree that `LinearFunction` is a very nice encapsulation! But two reservations:

* I do worry about the length of function names — e.g. when trying to get the slope for a linear fee, we would have mouthfuls like `(FeePolicy{linearFunctionIntercept=b,linearFunctionSlope=a})`. Typing such a thing has become less of an issue thanks to tab-completion, but it still hits the line length.
    But in this case, the encapsulation actually helps us: I think it would be perfect to call the fields simply `intercept` and `slope`. (This quickly becomes an issue of how we want to disambiguate record fields. I'm in favor of a "modules as namespaces" + (Maybe "`_` prefix to highlight fieldiness" or "`#` prefix to highlight fieldiness") approach.)
* Less importantly, I see no likely future in which a) `FeePolicy` will be a function different from linear or in which b) we need to do interesting computations on `LinearFunction`. However, even with this reservation, the benefit of meaningful encapsulation would be worth it in my opinion.